### PR TITLE
Add Vite support

### PR DESCRIPTION
**Pull request by mkreuzmayr**
I added Vite support via a plugin under `src/vite/index.ts`

The vite plugin can be used in two different ways, it can be set up manually or automatically with a wrapper function named `withKeycloakify` that automatically modifies the config accordingly. It was the best compromise I could find to reduce the boilerplate.

`vite.config.ts` using `withKeycloakify`:

```typescript
import react from '@vitejs/plugin-react';
import { defineConfig } from 'vite';
import { withKeycloakify } from 'keycloakify/vite';

// https://vitejs.dev/config/
export default defineConfig(
  withKeycloakify({
    plugins: [react()],
    assetsInclude: '**/*.md',
    server: {
      port: 3000,
    },
  })
);
```

`vite.config.ts` with manual setup:

```typescript
import react from '@vitejs/plugin-react';
import { defineConfig } from 'vite';
import { keycloakify, renderBuiltUrl } from 'keycloakify/vite';

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [react(), keycloakify()],
  define: {
    'process.env': {},
  },
  assetsInclude: '**/*.md',
  build: {
    outDir: 'build',
    assetsDir: 'static',
    target: 'es2015',
  },
  experimental: {
    renderBuiltUrl,
  },
  server: {
    port: 3000,
  },
});
```

EDIT:

What still needs to be done is to skip keycloakify build operations that have already been made by rollup.

### CODE REVIEW

File: src/vite/index.ts [hunk 1]
```diff
@@ -0,0 +1,92 @@
+import { load as cheerioLoad } from "cheerio";
+import { ftlValuesGlobalName } from "../bin/keycloakify/ftlValuesGlobalName";
+import { generateCssCodeToDefineGlobals, replaceImportsInCssCode } from "../bin/keycloakify/replacers/replaceImportsInCssCode";
+
+// need to be imported from vite/dist/node because of circular dependency error
+import type { RenderBuiltAssetUrl, UserConfig } from "vite/dist/node";
+
+export function withKeycloakify(config: UserConfig = {}): UserConfig {
```
Comment (willwill96) [inline src/vite/index.ts:8]: What do you think of removing the `withKeycloakify` helper & replacing it with a [config](https://vitejs.dev/guide/api-plugin.html#config) option on the plugin?

**Review by willwill96 — COMMENTED** [review]
Moving to vitejs will be a huge win! One cool thing I realized is that this might enable the ability for cross-framework development if you pair this with Astro.

**Comment by garronej** [comment]
Thank you guys, your rocks!  
I'll review that and get this merged and documented ASAP!

**Comment by garronej** [comment]
I just gave it a quick look.  

Correct me if I'm wrong but it seems like currently the only scenario supported is: a standalone keycloak theme ans the associated app if any must be hosted at the root of the domain.  
[`--external-assets`](https://github.com/keycloakify/keycloakify-starter/blob/main/README.md#standalone-vs---external-assets) isn't took into account, nor the path of the package.json homepage field (I don't know if it's used in Vite though).  
There are different case to consider:  
- External assets: the app and the keycloak server share the same domain: The app: https://starter.keycloakify.dev, the Keycloak server https://starter.keycloakify.dev/auth )
- External assets: the app and the ketcloak server are on different domain: The app: https://starter.keycloakify.dev, the Keycloak server https://auth.code.gouv.fr/auth, current scenario) 
- If the app is hosted in a sub path like https://starter.keycloakify.dev/a/b/c instead of https://starter.keycloakify.dev/

We would need to access the build option in the vite plugin:  
https://github.com/keycloakify/keycloakify/blob/fb98a9c38315a2a804cd7c6d803e3b9bec27a8a2/src/bin/keycloakify/BuildOptions.ts#L51-L86

**Comment by garronej** [comment]
I just gave you guys write access to the repo and the starter.  
@mkreuzmayr would you move this PR to a branch?  
This way we will be able to release a candidate juste by updating the package.json number to "7.2.0-rc.0" for example.

File: src/vite/index.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    config.plugins = [...(config.plugins ?? []), keycloakify()];
+
+    config.define = {
+        ...config.define,
+
+        // keycloakify expects a process.env becuase it looks for PUBLIC_URL
+        // we define the PUBLIC_URL as the base url of the vite config
+        "process.env": config.base ? { PUBLIC_URL: config.base } : {}
```
[the remark this answers is not in the document]
Comment (mkreuzmayr) [reply src/vite/index.ts:16]: Totally correct.

**Comment by mkreuzmayr** [comment]
For the future I think we should discuss where keycloakify wants to go, `withKeycloakify` was a compromise I made to integrate Vite without touching the current state of the keycloakify build pipeline.

I think there are two possible routes that keycloakify could go in the future.

## 1. Leave Vite and Keycloakify seperated

Leave keycloakify as it is and add the vite plugin as a preprocessor before running keycloakify. Then in keycloakify check if vite was used and skip the unnecessary build steps accordingly.

This requires vite as an external dependency. (additional setup by user is needed) Vite and keycloakify will be two seperated build operations that way. which would leave the possibility to use something other than vite to build the app.

This goes more the direction where you can combine your keycloak theme with your app in one project that may needs to have different build settings or a completely different build pipeline.


## 2. Integrate Vite into Keycloakify

Fully integrate vite into keycloakify and further leverage the transform funtions that vite/rollup provide. So we start the vite transformation programmatically within keycloakify and define a convention of settings that are used by default. So we have full control over the config and there will be less edge cases where custom Vite configs would create interference with keycloakify. 

The user only starts by installing keycloakify in a empty project and can add pages from there on without configuring vite. (Or clone a template) This would require a seperation of app and keycloak theme for some apps that need exotic configs. It would  also be simpler for the keycloakify team to release changes to the build process that would otherwise maybe destroy a specific build config of a user.


### Personal Opinion
I can only speak for me, but I do not see any value in having my app and keycloak theme in one project. With the rise and growing adoption of monorepo tools like turbo/NX in the web ecosystem I think there will not be many engineers that require the feature of having everything in one project, so I would rather focus on a standalone theme approach.

File: src/vite/index.ts (hunk 1 above)
Comment (mkreuzmayr) [reply src/vite/index.ts:8]: I thought about this, but we need other changes to the config besides adding the plugin.

File: src/vite/index.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+        // keycloakify expects build dir to be in /build
+        outDir: "build",
+
+        // keycloakify expects assets to be in /build/static
+        assetsDir: "static",
+
+        // most supported target
+        target: "es2015"
+    };
```
[the remark this answers is not in the document]
Comment (mkreuzmayr) [reply src/vite/index.ts:30]: https://github.com/keycloakify/keycloakify/pull/275#issuecomment-1482702358

**Comment by garronej** [comment]
@mkreuzmayr Thank you for articulating clearly what are the options moving forward, this is very useful.  

First I would like to clarify one thing.  
Theme being build with or without `--external-assets` and having the App and the Keycloak theme in the same project are two different things.  
I mean for `--external-assets` to make sense you **must** be in a setup where the App and the theme are shared but you can very well be in a setup where the app and the theme are shared without building with `--external-assets`.  

Now your argumentation made me think, why, for a start, don't we drop the `--external-assets` option altogether?  è
At first it was very important for performance since we didn't have lazy loading support like we have today.  
In consequence when you had a theme that was sharing the project with an app, the login page would make you download a large JavaScript bundle.  
`--external-assets` helped circumvent the problem by allowing you to enable resource caching so the bundle wouldn't be re-downloaded if it had been downloaded already on the main app.  
But this was a bandaid anyway. Now that everything is lazy loaded the .js is split in multiple little part and the amount of js that has to be downloaded for the login page isn't correlated anymore with the size of the main app.  
Beside this option, requires a lot of testing, explaining it to people is hard and it makes the build process much more complex since you have to have a single build for your theme and the App or you'd risk having a different hash (`main.xxxxxx.js`) and get 404s.  

So the first thing we could do is agreeing to drop `--external-assets` what do you think?

 > I can only speak for me, but I do not see any value in having my app and keycloak theme in one project. With the rise 
> and growing adoption of monorepo tools like turbo/NX in the web ecosystem I think there will not be many engineers 
> that require the feature of having everything in one project, so I would rather focus on a standalone theme approach.  

I get your point. Indeed, if you work in a monorepo setup it makes sense to set aside your main app and your keycloak theme since you can easily reference your UI components.  
Plus, for most of the projects, Next is a better option than a SPA framework. In 2023 we shouldn’t assume SPA to be the default option for a React App.   
But still. SPA have their use cases, I personally like being able to have my theme and my main app in the same place, it enables me to use the logic of my front end if I so wish.  
See for example here:  

https://user-images.githubusercontent.com/6702424/227660968-7a9e105e-4fd9-4eac-a88c-33c87dc63ea0.mov  

I make an API call to get the organizations available, I just had to import my core to do that. 
https://github.com/codegouvfr/sill-web/blob/cd032d04b188ec62e1a50d1e4a2a74013ad8d93f/src/ui/keycloak-theme/pages/shared/UserProfileFormFields.tsx#L43-L56  

All that to say, I like option 1 a lot and besides, I think that the way we currently replace the url with regexp is a bit hacky. Having a dedicated build process for the theme opens up the gate for a cleaner, more robust implementation. But I would like to hear what @willwill96 and @lordvlad think about it...

**Comment by lordvlad** [comment]
Great job! 

To be honest I don't have much experience with vite yet, so I can't comment much und the code. 

On the direction where this is going: I would love for keycloakify to be agnostic of the current build tool, cause I'm pretty sure next year already there will be a new shiny build tool/bundler/Framework. Or at the very least keep the integration as minimal as possible and work with configs and interfaces.

As for monorepo vs monolith I certainly see use cases for both, thou I agree that mono repo is the more flexible and powerful approach. If you need to use app methods in your theme you can factor them out into a lib. Monolith is attractive maybe in the beginning or when the scope is known to be limited, but as with any monolith it'll be prone to spaghetti. But I digress.

I see that you already made more directories configurable, I really need to hurry up with my branch which does same of the same + fixes dl caching 😅

**Comment by willwill96** [comment]
Heads up, some of these thoughts are a bit scattered, particularly towards the end.

This conversation got me thinking as well about what `keycloakify` is currently responsible for and this is what I have it broken down to. I'm new to the codebase though, so I might be missing or misinterpreting some things:
1. Transformation of index.html to .ftl & injection of `kcContext`
2. Bundling of everything into a jar
3. Providing KcContext API + React Library
4. Several utilities to make the asset paths from dev env -> prod work (replacers, `--external-assets`)

For me, the real value I see in `keycloakify` is in the first three points. I would put point 4 under the umbrella of "asset bundling" and it feels like these are primarily there to patch gaps of the current bundling framework (CRA/webpack). I have a few thoughts around these:
- With most modern frameworks I think, `--external-assets` is an unnecessary build option for keycloakify. I think even with dynamic import there is still value in the concept of it though, particularly if you want to serve assets behind a cdn. But between config options and build/environment variables it's relatively easy to get those setup in frameworks like Vite and Next. I wonder if a `--skipAssetTransformations` option along with docs on how to set it up in supported frameworks would be easier to maintain.
- The [CSS transformations](https://github.com/keycloakify/keycloakify/blob/main/src/bin/keycloakify/replacers/replaceImportsInCssCode.ts) into global css vars is a clever solution to a unique problem. It feels like this sort of thing should be supported in bundling frameworks, but I guess it is kind of a niche use case. The good thing is the current implementation already pretty framework agnostic. I think integrating the functionality within bundlers like is done as part of the vite plugin in this PR feels a bit cleaner, but leaving it as part of the `keycloakify` cli step would prevent more changes down the road to support more frameworks.
- The [JS transformations](https://github.com/keycloakify/keycloakify/blob/main/src/bin/keycloakify/replacers/replaceImportsFromStaticInJsCode.ts#L43-L67) that happen are pretty hacky. My understanding is that these are necessary because you don't know what the resource path is going to look like within keycloakify until runtime. Looking at some of the build outputs of a few major frameworks, it doesn't seem like there's that much consistency in how that transformation would work (even among different webpack-based frameworks). There may be a workaround that I'm missing, but I think this is where going for a purely framework agnostic approach for keycloakify breaks down. I think similar hacky solutions could be built for other build frameworks, but Vite is the only framework I've seen that offers the functionality directly from their API through `renderBuiltUrl`.

Ultimately, the more keycloakify can stay away from handling asset bundling concerns the better, because more focus can be given to the rest of the project. My personal preference would be to keep keycloakify as framework agnostic as possible, but I also feel like I'm coming to the conclusion that a lightweight preset/config per framework might be necessary if supporting multiple frameworks is important.

I think I my personal preference would be an approach somewhere in the middle of 1 & 2. I think I still see value in the standalone CLI of transforming html files into `ftl` & bundling everything up into the jar. But I think it might make maintenance much simpler if the keycloakify CLI could make the assumption that all the asset paths are already fixed up, all ideally handled by the bundler.

As for monolith vs standalone, I can see value in both approaches, but I think the more common use case is probably standalone. That being said, I think both monolith & standalone could be supported with either of the proposed approaches.

**Comment by garronej** [comment]
@willwill96, very well summed up.   

> With most modern frameworks I think, --external-assets is an unnecessary build option for keycloakify. I think even with dynamic import there is still value in the concept of it though, particularly if you want to serve assets behind a cdn. But between config options and build/environment variables it's relatively easy to get those setup in frameworks like Vite and Next. I wonder if a --skipAssetTransformations option along with docs on how to set it up in supported frameworks would be easier to maintain.  

💯 Ok great, let's drop `--external-assets` in v8. I like the idea of a `--skipAssetTransformations` option.  

> The [CSS transformations](https://github.com/keycloakify/keycloakify/blob/main/src/bin/keycloakify/replacers/replaceImportsInCssCode.ts) into global css vars is a clever solution to a unique problem. It feels like this sort of thing should be supported in bundling frameworks, but I guess it is kind of a niche use case. The good thing is the current implementation already pretty framework agnostic. I think integrating the functionality within bundlers like is done as part of the vite plugin in this PR feels a bit cleaner, but leaving it as part of the keycloakify cli step would prevent more changes down the road to support more frameworks.  

I feel leaving it to the Vite plugin feels **a lot** cleaner.  

> The [JS transformations](https://github.com/keycloakify/keycloakify/blob/main/src/bin/keycloakify/replacers/replaceImportsFromStaticInJsCode.ts#L43-L67) that happen are pretty hacky. My understanding is that these are necessary because you don't know what the resource path is going to look like within keycloakify until runtime. Looking at some of the build outputs of a few major frameworks, it doesn't seem like there's that much consistency in how that transformation would work (even among different webpack-based frameworks). There may be a workaround that I'm missing, but I think this is where going for a purely framework agnostic approach for keycloakify breaks down. I think similar hacky solutions could be built for other build frameworks, but Vite is the only framework I've seen that offers the functionality directly from their API through renderBuiltUrl.  

I hate this, I am embarrassed of this. I take the **output** of a webpack build and I use it as an **input**. Everything there is implementation specific, any patch update of Webpack can potentially beak Keycloakify. Plus, the search and replace are unpredictable.  

> Vite is the only framework I've seen that offers the functionality directly from their API through renderBuiltUrl  

Hell yeah, let's go with Vite, let's build from **source** instead of building from the output of Webpack.  
I mean this wouldn't mean that it would no longer be possible to set up Keycloakify in a CRA project or any other type of project. It would be like [vitest](https://vitest.dev/). Like in [this project](https://github.com/codegouvfr/react-dsfr) I have my tests that build with Vite and another build process for production.  

> But I also feel like I'm coming to the conclusion that a lightweight preset/config per framework might be necessary if supporting multiple frameworks is important.  

Correct me if I'm wrong but, like Vitest can be installed in any framework without needing specific integration to the meta framwork, it would be true for Keycloakify as well. We wouldn't need to integrate with a specific framework.  
The only thing that I see could be a pain is that not [all bundlers interprets the import of assets the same way](https://react-dsfr.etalab.studio/assets#rely-on-your-bundler). But I'm getting ahead of myself, this shouldn't be a major blocker.  

> As for monolith vs standalone, I can see value in both approaches, but I think the more common use case is probably standalone.  

The debate isn't monolith vs standalone. I think we all agree that The UI toolkit, the frontend and the backend should be different projects, linked together via a Monorepo framework or in some other way. That's not the point.  
The pain is: should we support the theme and the frontend code to be share the same project.  
And for me the answer is yes.  
The Keycloakify homepage says:  
*Wouldn't it be great if we could just design the login and register pages like any other pages of the app?*  
I think it's a powerful selling point, I like being able to reuse my fontend logic in Keycloak's user facing pages.  

> That being said, I think both monolith & standalone could be supported with either of the proposed approaches.  

That's the point. Implementing the first solution would mean that we wouldn't have anything to do to support this usecase.    

  
Well guys, in any case, I'm very exited about Keycloakify v8, I think all this has the potential to take the project to the next level.

**Comment by willwill96** [comment]
> Hell yeah, let's go with Vite, let's build from source instead of building from the output of Webpack.
I mean this wouldn't mean that it would no longer be possible to set up Keycloakify in a CRA project or any other type of project. It would be like [vitest](https://vitest.dev/). Like in [this project](https://github.com/codegouvfr/react-dsfr) I have my tests that build with Vite and another build process for production.

I was actually thinking much smaller scale & thinking of only supporting vite-based apps out of the box. This would mean if devs want to co-locate the front-end app & keycloak theme, Vite would be the only fully supported option. Theoretically if someone figured out a way to make dynamic asset paths work in other frameworks, they would be supported too, because the keycloakify CLI only expects an index.html & set of static assets. The javascript asset path transformations is the real blocker for making this work in most frameworks. 

Compiling from source, independent of framework is an interesting idea though. With that approach, it might be important to note, projects wouldn't actually be sharing assets between the frontend app and the keycloak templates, because the build systems would generate completely separate sets of assets. You mentioned this already, but there would also likely be some issues around asset imports. 

> The debate isn't monolith vs standalone. I think we all agree that The UI toolkit, the frontend and the backend should be different projects, linked together via a Monorepo framework or in some other way. That's not the point.
The pain is: should we support the theme and the frontend code to be share the same project.

If the keycloak theme and frontend code have separate build systems, is the only remaining benefit to co-locating them organizational/code-sharing? I don't want to push this too much, but I feel like if that's the case, it's easier to manage the independent build systems and share code between the two in a monorepo setup like the following:

```
| packages/core
| packages/front-end
//-->   depends on & imports code from packages/core
| packages/keycloak-theme
//-->   depends on & imports code from packages/core
```

I worry that keeping the build systems separate but co-located in the same project might introduce confusion when things go wrong (asset imports being handled differently, intellisense conflicts, failed uses of framework specific components or utilities (ex in nextjs, [NextLink](https://nextjs.org/docs/api-reference/next/link) or [next/config](https://nextjs.org/docs/api-reference/next.config.js/runtime-configuration)). I might be over-worrying about the possibility of this though.

**Comment by kasir-barati** [comment]
Just a question, I came here since my app uses Vite and I was wondered is it a good idea to develop my Signin and Signup pages inside the same monorepo. Thanks for answering this kinda off-topic question and bearing with me :heart:

**Comment by garronej** [comment]
Hi @kasir-barati,  
That's what we are curently argueing. I assume that when you say `inside the same monorepo` you mean you'll be [in this kind of scenario](https://github.com/codegouvfr/sill-web/tree/main/src) where your Keycloak theme is colocated with your fronted code.  
The answer is yes, this is how I do things and I see value doing things this way. So we will try do move in a direction that still enables it.  

However you are very early to the party, Vite isn't supported yet. @willwill96 and @mkreuzmayr got it working but they are experts.

**Comment by garronej** [comment]
> I worry that keeping the build systems separate but co-located in the same project might introduce confusion when things go wrong (asset imports being handled differently, intellisense conflicts, failed uses of framework specific components or utilities (ex in nextjs, [NextLink](https://nextjs.org/docs/api-reference/next/link) or [next/config](https://nextjs.org/docs/api-reference/next.config.js/runtime-configuration)). I might be over-worrying about the possibility of this though.  

It would have to be investigated but I already had do deal with this kind of things and I got away with helper functions. Like [this one](https://github.com/codegouvfr/react-dsfr/blob/main/src/tools/getAssetUrl.ts) that return the corect asset url regardless of if we are in Vite or Next.  
I think if we create a clear documentation section with:  
- Usage in colocation with Next.js: ...listing pitfalls to be aware of and workarounds ...
- Usage in colocation with CRA:      ...
- You are using asset import plugin: be aware that ... 

I think we could make this work. I realy would like to keep enabeling collocation, even if there is pitfalls.  
I think Vitest prooves that this approach works. (Thanks @GervinFung btw for introducting it in Denoify)  

> For me, as long as I have the ability to modify the vite.config file that keycloak is using under the hood, I will be a happy camper 😄  

😄😄  Yes, it would be a requirement that the `keycloakify.config.ts` would enables to extend the underlying Vite config.  

@mkreuzmayr I'd love to have your input on [this](https://github.com/keycloakify/keycloakify/pull/275#issuecomment-1483923565).

**Comment by garronej** [comment]
Thank for you answer @mkreuzmayr!  
I'm behind on my other project schedule so I don't have time to make a comprehensive response right now but: 

> I get your comparison with vitest but it is a bit off becuase we are somewhat dependent on the build output. Vitest is a standalone test suite running independently.    

Yes we are currently dependent of the output but why couldn't we built from the source using vite internally? It would be much cleaner don't you think?  
There would be whatever build process for the main App and completely independent build for the Keycloak theme.  
We would just need to know where the `index.txs` of the theme is.  
This would enable to colocate the theme with Next.js, Vite, CRA, whatever.   
Set aside the problem of the assets imports already mentioned I don't see why this approach couldn't work. (but this is very hypotetical, there's maybe a good reason that I don't see)

**Comment by mkreuzmayr** [comment]
> Yes we are currently dependent of webpack build output but why couldn't we built from the source using vite internally? Let's be ambitious, It would be much cleaner don't you think?

This is what I meant here.

> ## 2. Integrate Vite into Keycloakify
> So we start the vite transformation programmatically within keycloakify and define a convention of settings that are used by default. So we have full control over the config and there will be less edge cases where custom Vite configs would create interference with keycloakify.

---

> I'm behind on my other project schedule so I don't have time to make a comprehensive response right now but:

Same 🥴

**Comment by kasir-barati** [comment]
> Hi @kasir-barati, That's what we are curently argueing. I assume that when you say `inside the same monorepo` you mean you'll be [in this kind of scenario](https://github.com/codegouvfr/sill-web/tree/main/src) where your Keycloak theme is colocated with your fronted code. The answer is yes, this is how I do things and I see value doing things this way. So we will try do move in a direction that still enables it.
> 
> However you are very early to the party, Vite isn't supported yet. @willwill96 and @mkreuzmayr got it working but they are experts.

To be perfectly honest with you that's not how I thought about it. I was thinking more like this: that create a separate directory for it and initialize a new react app in it. But now that I am thinking about it I like to say why not like this: 
image

It is much easier to ship code to the prod since we will have separate code bases and separate build. But at the same time we can share components and libraries.

In my [you-say app](https://github.com/kasir-barati/you-say) I would have something like this: 
| - you-say
| -- apps
| ---- frontend
| ---- backend
| -- packages
| ---- loginout

I assumed that I can share common 3rd party libs and components between my frontend app and my loginout package. Can you see any drawback?

**Comment by garronej** [comment]
Hey @kasir-barati,  
Could you move your question to the discution section?  
I want would like this PR discution to stay precisely on topic.  

Thanks

**Comment by garronej** [comment]
Storybook is also an interesting study case.  
They internalize the build système and thus enable to integrate in any app.  
They are moving to vite as well: https://storybook.js.org/blog/first-class-vite-support-in-storybook/

**Comment by garronej** [comment]
@all-contributors please add @willwill96

**Comment by garronej** [comment]
@all-contributors please add @mkreuzmayr

**Comment by garronej** [comment]
@all-contributors please add @lordvlad

**Comment by garronej** [comment]
@all-contributors please add @willwill96 for tests and code

**Comment by garronej** [comment]
@all-contributors please add @lordvlad for tests and code

**Comment by garronej** [comment]
@all-contributors please add @mkreuzmayr  for tests and code

**Comment by garronej** [comment]
@all-contributors please add @kasir-barati for doc

**Comment by garronej** [comment]
Damn sorry all.  
I entered the wrong command and literally overwrote `main`.  
Thanks got I had a local copy but it closed this PR and I have no choice but open another one :(

**Comment by garronej** [comment]
Lesson learned: Don't copy paste GPT4 generated commands blindly, however convincing they may be.

**Comment by garronej** [comment]
And don't add `-f` if you left your brain on the desk

**Pull request closed without merging**